src/storage/src/render/sources.rs

```diff
@@ -366,37 +367,88 @@ where
+                                        };
+                                    let (stream, tok) = persist_source::persist_source_core(
+                                        scope,
+                                        id,
+                                        persist_clients,
+                                        description.ingestion_metadata,
+                                        Some(as_of),
+                                        Antichain::new(),
+                                        None,
+                                        None,
+                                        granular_flow_control,
+                                        // Copy the logic in DeltaJoin/Get/Join to start.
+                                        |_timer, count| count > 1_000_000,
+                                    );
+                                    (stream.as_collection(), Some(tok), feedback_handle)
+                                } else {
+                                    (Collection::new(empty(scope)), None, None)
+                                };
+                            let (upsert, health_update) = crate::render::upsert::upsert(
+                                &upsert_input.enter(scope),
+                                upsert_envelope.clone(),
+                                refine_antichain(&resume_upper),
+                                previous,
+                                previous_token,
+                                base_source_config,
+                                &storage_state.instance_context,
+                                &storage_state.dataflow_parameters,
+                            );
+
+                            // If granular backpressure is enabled, we probe the upsert operator's
+                            // output, which is the easiest way to extract frontier information.
+                            let upsert = match feedback_handle {
+                                Some(feedback_handle) => {
+                                    use mz_timely_util::probe::ProbeNotify;
+                                    let handle = mz_timely_util::probe::Handle::default();
+                                    let upsert =
+                                        upsert.inner.probe_notify_with(vec![handle.clone()]);
+                                    mz_timely_util::probe::source(
+                                        scope.clone(),
+                                        format!("upsert_probe({id})"),
+                                        handle,
+                                    )
+                                    .connect_loop(feedback_handle);
+                                    upsert
+                                }
+                                None => upsert.inner,
+                            };
+
+                            (
+                                upsert.leave().map(|(d, t, r)| (d, t.0, r)),
```

Comment:
We should turn `upsert` into a DD `Collection` because `Collection::leave` does this automatically

Reply:
lmao i missed this cause I looked at the `docs.rs` dd docs LMAO